contracts/extensions/Auctions/DutchAuction.sol

```diff
@@ -0,0 +1,483 @@
+
+    emit PlaceBid(msg.sender, collection, item, msg.value, _memo);
+  }
+
+  /**
+   * @notice Settles the auction after expiration if no valid bids were received by sending the item back to the seller. If a valid bid matches the current price at the time of settle call, the item is sent to the bidder. Proceeds will be distributed separately by calling `distributeProceeds`.
+   *
+   * @param collection ERC721 contract.
+   * @param item Token id to settle.
+   */
+  function settle(
+    IERC721 collection,
+    uint256 item,
+    string calldata _memo
+  ) external override nonReentrant {
+    bytes32 auctionId = keccak256(abi.encodePacked(collection, item));
+    DutchAuctionData memory auctionDetails = auctions[auctionId];
+
+    if (auctionDetails.info == 0) {
+      revert INVALID_AUCTION();
+    }
+
+    uint256 lastBidAmount = uint256(uint96(auctionDetails.bid >> 160));
+    uint256 minSettlePrice = currentPrice(collection, item);
+
+    if (minSettlePrice < lastBidAmount) {
+      address buyer = address(uint160(auctionDetails.bid));
+
+      collection.transferFrom(address(this), buyer, item);
+
+      emit ConcludeAuction(
+        address(uint160(auctionDetails.info)),
+        buyer,
+        collection,
+        item,
+        lastBidAmount,
+        _memo
+      );
+    } else {
+      uint256 expiration = uint256(uint64(auctionDetails.prices >> 192));
+      if (block.timestamp > deploymentOffset + expiration) {
+        // NOTE: return token back to seller after auction expiration if highest bid was below settlement price
+        collection.transferFrom(address(this), address(uint160(auctionDetails.info)), item);
+
+        delete auctions[auctionId];
+        delete auctionSplits[auctionId];
+
+        emit ConcludeAuction(
+          address(uint160(auctionDetails.info)),
+          address(0),
+          collection,
+          item,
+          0,
+          _memo
+        );
+      }
+    }
+  }
+
+  /**
+   * @notice This trustless method removes the burden of distributing auction proceeds to the seller-configured splits from the buyer (or anyone else) calling settle().
+   */
+  function distributeProceeds(IERC721 _collection, uint256 _item) external override {
+    bytes32 auctionId = keccak256(abi.encodePacked(_collection, _item));
+    DutchAuctionData memory auctionDetails = auctions[auctionId];
+
+    if (auctionDetails.info == 0) {
+      revert INVALID_AUCTION();
+    }
+
+    uint256 lastBidAmount = uint256(uint96(auctionDetails.bid >> 160));
+    uint256 minSettlePrice = currentPrice(_collection, _item);
+
+    if (minSettlePrice > lastBidAmount) {
+      // nothing to distribute no valid bid
+      revert INVALID_PRICE();
+    }
+
+    if (_collection.ownerOf(_item) == address(this)) {
+      // proceeds can be collected, but we still own the token, send it to the bidder
+      address buyer = address(uint160(auctionDetails.bid));
+      _collection.transferFrom(address(this), buyer, _item);
```

Comment:
Reentrency, malicious actor could make a NFT, auction it, place a bid themselves, then reenter using this transferFrom to drain the contract

Reply:
would adding `nonReentrant` be sufficient here?

Reply:
yeah that works